pkg/sql/gcjob_test/gc_job_test.go

```diff
@@ -506,6 +508,54 @@ func TestGCTenant(t *testing.T) {
 	})
 }
 
+// This test exercises code whereby an index GC job is running, and, in the
+// meantime, the descriptor is removed. We want to ensure that the GC job
+// finishes without an error.
+func TestDropIndexWithDroppedDescriptor(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	gcJobID := make(chan jobspb.JobID)
+	s, sqlDB, kvDB := serverutils.StartServer(t, base.TestServerArgs{
+		Knobs: base.TestingKnobs{
+			JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
+			GCJob: &sql.GCJobTestingKnobs{RunBeforeResume: func(jobID jobspb.JobID) error {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case gcJobID <- jobID:
+					return nil
+				}
+			}},
+		},
+	})
+	defer s.Stopper().Stop(ctx)
+	defer cancel()
+	tdb := sqlutils.MakeSQLRunner(sqlDB)
+	tdb.Exec(t, "CREATE TABLE foo (i INT PRIMARY KEY, j INT, INDEX(j, i))")
+	var tableID catid.DescID
+	var indexID catid.IndexID
+	tdb.QueryRow(t, `
+SELECT descriptor_id, index_id
+  FROM crdb_internal.table_indexes
+ WHERE descriptor_name = 'foo'
+   AND index_name = 'foo_j_i_idx';`).Scan(&tableID, &indexID)
+	tdb.Exec(t, "DROP INDEX foo@foo_j_i_idx")
+	codec := s.ExecutorConfig().(sql.ExecutorConfig).Codec
+	errCh := make(chan error, 1)
+	// Allow the job to proceed in parallel to deleting the descriptor.
+	jobID := <-gcJobID
+	go func() {
+		k := catalogkeys.MakeDescMetadataKey(codec, tableID)
+		_, err := kvDB.Del(ctx, k)
+		errCh <- err
+	}()
```

Comment:
Is the ordering of those two steps reversed?
The `DROP INDEX` stmt above will ensure we stall the index-gc job (due to the testing knob). We want to test the scenario where the the index-gc job failed to find the table descriptor, so shouldn't we drop the table first before we allow the index-gc job to proceed?

Reply:
There's no guarantee of ordering between the two. I don't think it changes the test to say `go` first, but I'll do it to clarify intention.

Reply:
I still didn't understand this one: gcJobID is a 0-capacity channel. so `jobID := <- jcJobID` will unblock the `DROP INDEX` first and then we schedule a go routine that will delete the descriptor from `system.descriptor` table, which makes the `DROP INDEX` gc-job may or may not observe the deletion.

Shouldn't we make sure that we deleted the descriptor before we "let go" of the `DROP INDEX` index-gc job, which ensures that the gc-index job will observe the deletion?

Reply:
If this test were better, it'd intercept behavior of the gc job and delete the descriptor at specific moments, but I didn't feel that it was worth it, so I just ran the test on stress for a few minutes and called it a day.

Reply:
I made the test better.